python/orca/tutorial/tensorflow/ncf_orca.py

```diff
@@ -0,0 +1,155 @@
+                           conf={"spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.driver.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.options.claimName": "nfsvolumeclaim",
+                                 "spark.kubernetes.executor.volumes.persistentVolumeClaim.nfsvolumeclaim.mount.path": "/bigdl2.0/data"})
+elif cluster_mode == "spark-submit":  # To test k8s using spark-submit
+    sc = init_orca_context(cluster_mode="spark-submit")
+
+
+data_path = "./bigdl2.0/data"
+data_type = "ml-1m"
+# Need spark3 to support delimiter with more than one character.
+spark= OrcaContext.get_spark_session()
+from pyspark.sql.types import StructField, StructType, IntegerType
+
+schema = StructType(
+    [
+        StructField('user', IntegerType(), True),
+        StructField('item', IntegerType(), True)
+    ]
+)
+df = spark.read.csv("{}/{}/ratings.csv".format(data_path, data_type), sep="::",schema=schema,
+                     header=False)
+min_user_id = df.agg({"user": "min"}).collect()[0]["min(user)"]
+max_user_id = df.agg({"user": "max"}).collect()[0]["max(user)"]
+min_item_id = df.agg({"item": "min"}).collect()[0]["min(item)"]
+max_item_id = df.agg({"item": "max"}).collect()[0]["max(item)"]
+print(min_user_id, max_user_id, min_item_id, max_item_id)
+from pyspark.sql import functions
+df = df.withColumn('label', functions.lit(1))
+
+from pyspark.sql import functions as F
+from pyspark.sql import types as T
+import random
+
+neg_scale = 4
+def neg_sample(x):
+    item_count = len(x) * neg_scale
+    max_count = max_item_id - len(set(x))
+    neg_item = random.sample(set(range(min_item_id, max_item_id+1)) - set(x), min(item_count,max_count))
+    return neg_item
+
+neg_sample_udf = F.udf(neg_sample, T.ArrayType(IntegerType(), False))
+
+df_neg= df.groupBy('user').agg(neg_sample_udf(F.collect_list('item')).alias('item_list'))
+from pyspark.sql.functions import *
+df_neg = df_neg.select(df_neg.user, explode(df_neg.item_list))
+df_neg = df_neg.withColumn('label', functions.lit(0))
+df = df.unionAll(df_neg)
+num_sample = df.count()
+train_df, test_df = df.randomSplit([0.8, 0.2],100)
+
+
+batch_size=256
+epochs=5
+model_dir='./'
+
+# create an Estimator
+backend = 'spark'
+# est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1, backend=backend)
+est = Estimator.from_keras(model_creator=model_creator, workers_per_node=1)
+
+stats = est.fit(train_df,
+                epochs=epochs,
+                batch_size=batch_size,
+                feature_cols=['user', 'item'],
+                label_cols=['label'],
+                steps_per_epoch=int(60000 // batch_size),
+                validation_data=test_df,
+                validation_steps =int(10000 // batch_size))
```

Comment:
should not hard code 60000 and 10000, use df.count

Reply:
> should not hard code 60000 and 10000, use df.count

When step is set to 60000/10000 (refer to tutorial), the validation accuracy is ideal. When df.count is used, the result is abnormal.

Reply:
Then you need to figure out why it is abnormal instead of hard code the values.